SSA/Projects/CSE/CSE.lean

```diff
@@ -0,0 +1,539 @@
+        ⟨.lete e' body',  by
+            intros VΓ
+            simp [Com.denote]
+            simp [Lets.denote_lete] at hbody' ⊢
+            rw [← hbody']
+            rw [he']
+            done⟩
+      | .some ⟨v', hv'⟩ =>
+        let s' := s.snocOldExpr2Cache (enew := e') (eold := e) (henew := by { intros V; rw [he'] })
+          (vold := v') (hv := by {intros V; rw [hv'] }) -- add this expression into the cache for the latest variable.
+        let ⟨body', hbody'⟩ := s'.cseCom body
+        -- TODO: delete the ``e` to get a `body'` in context `Γ`, not `Γ.snoc α`.
+        ⟨.lete e body' -- we still keep the `e` for now. In the next version, we will delete the `e`
+        , by
+            intros V
+            simp [Com.denote]
+            simp [Lets.denote_lete] at hbody' ⊢
+            specialize (hbody' V)
+            specialize (he' V)
+            rw [he'] at hbody'
+            apply hbody'
+            done
+        ⟩
+
+end -- mutual.
+
+/-- common subexpression elimination entry point. -/
+unsafe def cse' [DecidableEq Ty] [DecidableEq Op] [OpSignature Op Ty] [Goedel Ty] [OpDenote Op Ty]
+  {α : Ty} {Γ : Ctxt Ty} (com: Com Op Γ α) :
+  { com' : Com Op Γ α // ∀ (V: Ctxt.Valuation Γ), com.denote V = com'.denote V } :=
+    let ⟨com', hcom'⟩ := State.cseCom (State.empty Lets.nil) com
+    ⟨com', by {
+      intros V
+      specialize (hcom' V)
+      simp [Lets.denote] at hcom'
+      assumption
+    }⟩
+
+namespace Examples
+/-- A very simple type universe. -/
+inductive ExTy
+  | nat
+  | bool
+  deriving DecidableEq, Repr
+
+@[reducible]
+instance : Goedel ExTy where
+  toType
+    | .nat => Nat
+    | .bool => Bool
+
+inductive ExOp :  Type
+  | add : ExOp
+  | beq : ExOp
+  | cst : ℕ → ExOp
+  deriving DecidableEq, Repr
+
+instance : OpSignature ExOp ExTy where
+  signature
+    | .add    => ⟨[.nat, .nat], [], .nat⟩
+    | .beq    => ⟨[.nat, .nat], [], .bool⟩
+    | .cst _  => ⟨[], [], .nat⟩
+
+@[reducible]
+instance : OpDenote ExOp ExTy where
+  denote
+    | .cst n, _, _ => n
+    | .add, .cons (a : Nat) (.cons b .nil), _ => a + b
+    | .beq, .cons (a : Nat) (.cons b .nil), _ => a == b
+
+def cst {Γ : Ctxt _} (n : ℕ) : Expr ExOp Γ .nat  :=
+  Expr.mk
+    (op := .cst n)
+    (ty_eq := rfl)
+    (args := .nil)
+    (regArgs := .nil)
+
+def add {Γ : Ctxt _} (e₁ e₂ : Ctxt.Var Γ .nat) : Expr ExOp Γ .nat :=
+  Expr.mk
+    (op := .add)
+    (ty_eq := rfl)
+    (args := .cons e₁ <| .cons e₂ .nil)
+    (regArgs := .nil)
+
+attribute [local simp] Ctxt.snoc
+
+def ex1_pre_cse : Com ExOp ∅ .nat :=
+  Com.lete (cst 1) <|
+  Com.lete (cst 1) <|
+  Com.lete (add ⟨0, by simp⟩ ⟨1, by simp⟩) <|
+  Com.ret ⟨0, by simp [Ctxt.snoc]⟩
+#eval ex1_pre_cse
+
+unsafe def ex1_post_cse : Com ExOp ∅ .nat := (cse' ex1_pre_cse).val
+#eval ex1_post_cse
+/-
+CSE.Examples.ExOp.cst 1[[]] -- %1
+CSE.Examples.ExOp.cst 1[[]] -- %0
+CSE.Examples.ExOp.add[[%1, ,, %1]] -- see that the more recent use is dead.
+return %0
+-/
```

Comment:
We had such nice MLIR syntax. Maybe one test case with MLIR syntax would be great.

Also, can we actually write theorem that states input = dce input. This should be equal by reflection at least? Also, did you actually proof this correct?

Reply:
@tobiasgrosser We lost the MLIR syntax when we switched to `ICom` :)

Reply:
@tobiasgrosser We know that `input = DCE input`, that's the second component that's returned by DCE. I extract out the val, but yes, I can also grab the property.

Reply:
Yes, it's all verified correct.

Reply:
How cool. I would still like to see MLIR test cases, but maybe this can wait for after the deadline. I merge this for now. We should mention it in the paper!

Also, congrats for being side-tracked incredibly productively.